plugins/http_plugin/common.hpp

```diff
@@ -0,0 +1,274 @@
+        return std::make_shared<in_flight<T>>(std::forward<T>(object), plugin_state);
+    }
+
+    /**
+     * Construct a lambda appropriate for url_response_callback that will
+     * JSON-stringify the provided response
+     *
+     * @param thread_pool - pointer to the thread pool 
+     * @param plugin_state = reference to plugin state object
+     * @param session_ptr - pointer to the session object on which to invoke send_response
+     * @return lambda suitable for url_response_callback
+     */
+    auto make_http_response_handler(std::shared_ptr<eosio::chain::named_thread_pool> thread_pool, std::shared_ptr<http_plugin_state> plugin_state, detail::abstract_conn_ptr session_ptr) {
+    return [thread_pool, plugin_state, session_ptr]( int code, std::optional<fc::variant> response )
+         {
+            auto tracked_response = make_in_flight(std::move(response), *plugin_state);
+            if (!session_ptr->verify_max_bytes_in_flight()) {
+                return;
+            }
+
+            // post  back to an HTTP thread to to allow the response handler to be called from any thread
+            boost::asio::post(  thread_pool->get_executor(),
+                                [plugin_state, session_ptr, code, tracked_response=std::move(tracked_response)]() {
+                try {
+                    if( tracked_response->obj().has_value() ) {
+                        std::string json = fc::json::to_string( *tracked_response->obj(), fc::time_point::now() + plugin_state->max_response_time );
+                        auto tracked_json = make_in_flight( std::move( json ), *plugin_state );
+                        session_ptr->send_response( std::move( tracked_json->obj() ), code );
+                    } else {
+                        session_ptr->send_response( {}, code );
+                    }
+                } catch( ... ) {
+                    session_ptr->handle_exception();
+                }
+            });
+        };// end lambda
+    }
+
+    bool host_port_is_valid( const http_plugin_state& plugin_state, 
+                             const std::string& header_host_port, 
+                             const string& endpoint_local_host_port ) {
+        return !plugin_state.validate_host 
+                || header_host_port == endpoint_local_host_port 
+                || plugin_state.valid_hosts.find(header_host_port) != plugin_state.valid_hosts.end();
+    }
+
+    bool host_is_valid( const http_plugin_state& plugin_state, 
+                        const std::string& host, 
+                        const string& endpoint_local_host_port, 
+                        bool secure) {
+        if (!plugin_state.validate_host) {
+            return true;
+        }
+
+        // normalise the incoming host so that it always has the explicit port
+        static auto has_port_expr = std::regex("[^:]:[0-9]+$"); /// ends in :<number> without a preceeding colon which implies ipv6
```

Comment:
I think this would be better than regex: https://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/ip__address.html

Reply:
Was http_plugin designed to support IP v6?   If not, I can convert the addresses to IP v4 for comparison.  The initialization code  `http_plugin::plugin_initialize()` seems to assume IP v4, because its looking for a host/port and calling `resolve` with `tcp::v4()` as an argument.

Reply:
It probably was not designed to support IP v6, but if you can add support as part of this PR that would be a plus.

Specifically, instead of regex, you should be able to do something like this:

```
auto add = ip::address::make_address( host );
if( add.is_v6() ) 
```

Reply:
This was my thought as well.  We can just normalize addresses to IP v6, shouldn't require conditionals.

Reply:
Nice

Reply:
I have added support for IPv6 and reworked the allow_hosts logic.  The existing implementation struck me as a bit type-unsafe and did some unnecessary conversion between addresses and strings.  Hopefully it is cleaned up now.  The one thing we might want to take a closer look at is a function I added called get_ipaddr_host_port(), which takes either an IPv4 or IPv6 address string and splits it up into host/port, autodetecting which version it is.  I couldn't find a library implementation for this,  so I made a simple implementation to do this which assumes IPv6 addresses contain at least open bracket '[, though it doesn't have to be a closed bracket unless it contains a port number.  I believe there is an RFC which this should be the case for IPv6 addresses which contain port numbers, but I'm not sure about IPv6 addresses which do not contain port numbers..  

Also not sure how strict it should be, such as detecting the ':' separator (which wasn't done in current implementation even for IPv4)